doc/src/conf.py

```diff
@@ -113,7 +113,7 @@
 # Myst configuration (for .md files). See
 # https://myst-parser.readthedocs.io/en/latest/syntax/optional.html
 myst_enable_extensions = ["dollarmath", "linkify"]
-myst_heading_anchors = 2
+myst_heading_anchors = 6
```

Comment:
What does this do?

Reply:
The auto-creates anchors ("slugs") for headings (in Markdown which become HTML headings such as h6) down to the given level. I changed it from `h2` to `h6` so we can link to a heading of any level. As noted in #23777, I found that change [did not have any adverse effects on building the docs, and the Executable Books Project lead developer wrote that it "should be all good".](https://github.com/sympy/sympy/pull/23777#issuecomment-1221322927)

Reply:
I guess I'm just not clear what difference it makes. The deepest Markdown heading in the docs is level 4 in the custom functions guide and you seem to be able to link to those headings (e.g., https://docs.sympy.org/latest/guides/custom-functions.html#doit). 

What is the default value of this? Can we just leave it undefined in the conf file?

Reply:
The difference is whether you can cross-link to the heading _from markdown links_. You can always manually link to a heading, as in the link you provide above.

As a test, I inserted a link to an `h3` and it fails if `myst_heading_anchors = 2` or is not in the conf file, and works if `myst_heading_anchors = 6`. The default seems to be, in effect, zero because links to even the h1 heading fail if myst_heading_anchors is not in the conf file.

From the [MyST docs](https://myst-parser.readthedocs.io/en/latest/syntax/optional.html#auto-generated-header-anchors):

> To achieve this, use the myst_heading_anchors = DEPTH configuration option, where DEPTH is the depth of header levels for which you wish to generate links.
> 
> For example, the following configuration in conf.py tells the myst_parser to generate labels for heading anchors for h1, h2, and h3 level headings (corresponding to #, ##, and ### in markdown).
> 
> `myst_heading_anchors = 3`
> 
> You can then insert markdown links directly to anchors that are generated from your header titles in your documentation.

Reply:
I see. We don't even have that many markdown documents, so I'm not surprised this makes not difference to the build time. And even if it did, it's worth having in my opinion. 

Although I will say that it's usually a good idea to use the `(xref)=` syntax to manually create a cross reference because you can easily make those globally unique without having to worry about them being clobbered by something else in the future, and you can change the name of the header without breaking anything (although I guess in either case the docs build errors would alert you to the issue if that happened).

Reply:
Good deal.

> Although I will say that it's usually a good idea to use the (xref)= syntax to manually create a cross reference because you can easily make those globally unique without having to worry about them being clobbered by something else in the future

Do you mean by using some naming scheme to ensure the `xref` is unique? One of the benefits of the slugs (auto-creating anchors from headings) is that, because they're anchors, there's no possibility of conflict by putting the same heading name in two files because you have to preface the anchor with the file name, e.g. if page1 and page2 both have a heading `solve`, the links to them (from a third page) will be page1#solve and page2#solve, respectively. If you have the same heading twice in a file, MyST recognizes that and numbers the second heading.

I agree that downsides of these slug anchors are
- they can be broken by renaming the headings
- if referencing from another page, you have to know where the target page is in the hierarchy, e.g. ../../target.md, and that could change if you rearrange the hierarchy

At the same time, it does seem kind of arbitrary and difficult to track to add `xref`s to various documents. It also leads to pull requests that modify a (target) file when they really don't change the (displayed) content, just add an `xref` to it.

So each has upsides and downsides. I think slugs are good for references within the same document because you'll know if you change the heading name, thus that the slug name needs to change. Also, at least within Visual Studio Code, if you start a Markdown link by typing `#` (after an opening parenthesis or documentation file name), VS Code will suggest those slugs via autocomplete so you can just click on the one you want.

Reply:
Somewhat related, I just learned about [Markdown Link Check GitHub Action](https://github.com/gaurav-nelson/github-action-markdown-link-check). I think we already have this covered in the Sphinx test that runs automatically for a pull request; just mentioning it in case it's useful.

Reply:
Yes, typically I use a naming scheme. But maybe anchors are better (the xrefs are still useful though because they can refer to an arbitrary location in the document).

> Somewhat related, I just learned about [Markdown Link Check GitHub Action](https://github.com/gaurav-nelson/github-action-markdown-link-check). I think we already have this covered in the Sphinx test that runs automatically for a pull request; just mentioning it in case it's useful.

It shouldn't be necessary with MyST. Unless I'm mistaken, the Sphinx build will fail if a cross-reference can't be found. We could use `make linkcheck` to check external links (although that obviously has the potential to be flaky).